Output the storage layout of a contract via storageLayout artifact (#7589)
ethereum/solidity: docs/miscellaneous.rst

Layout of State Variables in Storage
************************************

> .. _storage-inplace-encoding:

Reviewer: Did you test these? As far as I know, they have to come before the header they refer to

Author: Yes, that works. I meant to refer to that specific paragraph and not the entire section.

JSON Output
===========

The storage layout of a contract can be requested via the :ref:`standard JSON interface <compiler-api>`.  The output is a JSON object containing two keys,
``storage`` and ``types``.  The ``storage`` object is an array where each
element has the following form:

.. code::

    {
        "ast_id":"2",
        "contract":"A",
        "label":"x",
        "offset":"0",
        "slot":"0",
        "type":"t_uint256"
    }

where the example above is the storage layout of ``contract A { uint x; }``
and

- ``ast_id`` is the id of the AST node of the state variable
- ``contract`` is the name of the contract
- ``label`` is the name of the state variable
- ``offset`` is the offset within the storage slot according to the encoding
- ``slot`` is the storage slot where the state variable resides
- ``type`` is an identifier used as key to the variable's type information

The given ``type``, in this case ``t_uint256`` represents an element in
``types``, which has the form:

.. code::

    {
        "encoding":"inplace",
        "label":"uint256",
        "numberOfBytes":"32",
        "numberOfSlots":"1"
    }

where

> - ``encoding`` tells how the data is stored

Reviewer 2: I'd say this is not very helpful and has to be expanded - it should list all possible values and what they mean exactly...

Author: This should be in the section right above this one that explains how the storage layout/encoding works.

Author: I can list the options linking them back to the section before.

Reviewer 2: Yeah, an exhaustive list of possible values linking to the respective session should be enough, if it's sufficiently explained there.

Author: Updated

- ``label`` is the canonical type name
- ``numberOfBytes`` the number of bytes that this type occupies within a slot
- ``numberOfSlots`` the number of slots that this type requires in storage

> Some types have extra information besides the four above. Mappings contain

Reviewer 2: You should probably list all of this, so including arrays with ``base`` and relate that to the information for ``encoding`` above and specify every additional field separately... I mean it's not like "some types have something extra", but that there are two kinds of fields: general fields and fields specific to the concrete encoding given by the value of ``encoding``, right?

I mean we could postpone all this for later, but it should actually be fully documented and not with an implicit "just try and figure it out" ;-).

Author: These are all the extra ones.

Author: I'm not sure I understand how this relates to the `encoding` field. Mapping and Arrays have extra keys and different encoding, but Structs have extra keys but the encoding is the same as value types.

Reviewer 2: I somehow missed the fact that ``base`` is in fact in the next line, sorry... and true about encoding not actually relating - still: I wonder, if there *should* be something relating to it, like a ``kind`` field being ``value``, ``tuple``/``struct``, ``array`` or ``mapping``... having to look for the presence of any of the extra keys to tell what it is, is not nice, is it?

Reviewer 2: But yeah - maybe I should read through the whole discussion about this before making random claims :-D.

Reviewer 2: I was thinking about e.g. what happens if I forgot that the ``members`` key might exist (or more realistically we add another sub-type in the future)... then I'd just consider the thing as a regular value type... but actually this is just fine... and may be for all imaginable such cases... Yeah, ok, fine ;-).

its ``key`` and ``value`` types, arrays have its ``base`` type, and structs

> list its ``members`` similarly to key ``storage``.

Reviewer 2: ```suggestion
list its ``members`` in the same format as the top-level ``storage`` object (see :ref:[head of section]).
```
?

Author: Updated

.. note::
  Handling invalidly encoded slots is currently not supported but may be added in the future.

JSON Output
===========

.. _storage-layout-top-level:

The storage layout of a contract can be requested via the :ref:`standard JSON interface <compiler-api>`.  The output is a JSON object containing two keys,
``storage`` and ``types``.  The ``storage`` object is an array where each
element has the following form:

.. code::

    {

>         "astId":"2",

Reviewer: It is a string, not a number?

Author: Updated

Reviewer 2: I'd actually prefer a string here... not every imaginable AST has numbered nodes, so this could actually be any kind of identifier... and I think it doesn't hurt for this output to be as generic as possible - but I'm fine with a number as well...

Reviewer: ok!

Reviewer: But now the code uses integers.

Author: Yes, changed all back to integers except `slot`, and added comment about `slot` in the documentation.

        "contract":"A",
        "label":"x",

>         "offset":"0",

Reviewer: String, not number?

Author: Updated

>         "slot":"0",

Reviewer: Here it might make sense to use a string because json numbers are probably not large enough.

Author: Yea, that's why I went with string for all

Reviewer: Maybe mention somewhere that it is a string? Or is there no real place?

Reviewer: And also mention that it can be very large.

Author: Mentioned it in the first part of the doc

        "type":"t_uint256"
    }

where the example above is the storage layout of ``contract A { uint x; }``
and

> - ``astId`` is the id of the AST node of the state variable

Reviewer: ```suggestion
- ``astId`` is the id of the AST node of the state variable's declaration
```

Author: Updated

> - ``contract`` is the name of the contract

Reviewer: without path prefix?

- ``label`` is the name of the state variable

> - ``offset`` is the offset within the storage slot according to the encoding

Reviewer: ```suggestion
- ``offset`` is the offset in bytes within the storage slot according to the encoding
```

Author: Updated

> - ``slot`` is the storage slot where the state variable resides

Reviewer: ```suggestion
- ``slot`` is the storage slot where the state variable resides or starts
```

Author: Updated

> - ``type`` is an identifier used as key to the variable's type information

Reviewer: ```suggestion
- ``type`` is an identifier used as key to the variable's type information (described in the following)
```

Author: Updated

The given ``type``, in this case ``t_uint256`` represents an element in
``types``, which has the form:

.. code::

    {
        "encoding":"inplace",
        "label":"uint256",
        "numberOfBytes":"32",
    }

where

- ``encoding`` how the data is encoded in storage, where the possible values are:

  - ``inplace``: data is laid out contiguously in storage (see :ref:`above <storage-inplace-encoding>`).
  - ``mapping``: Keccak-256 hash-based method (see :ref:`above <storage-hashed-encoding>`).
  - ``dynamic_array``: Keccak-256 hash-based method (see :ref:`above <storage-hashed-encoding>`).

>   - ``string``: single slot or Keccak-256 hash-based depending on the data size (see :ref:`above <bytes-and-string>`).

Reviewer: Historically (and in the code-base) `bytes` is the generic term, but I'm fine with either

- ``label`` is the canonical type name.
- ``numberOfBytes`` is the number of used bytes. Note that if ``numberOfBytes > 32`` this means that more than one slot is used.

Some types have extra information besides the four above. Mappings contain

> its ``key`` and ``value`` types, arrays have its ``base`` type, and structs

Reviewer: ```suggestion
its ``key`` and ``value`` types (again referencing an entry in this mapping of types), arrays have its ``base`` type, and structs
```

Author: Updated

> list its ``members`` in the same format as the top-level ``storage``

Reviewer: ```suggestion
list their ``members`` in the same format as the top-level ``storage``
```

Author: Updated

.. note::
  Handling invalidly encoded slots is currently not supported but may be added in the future.

JSON Output
===========

.. _storage-layout-top-level:

The storage layout of a contract can be requested via the :ref:`standard JSON interface <compiler-api>`.  The output is a JSON object containing two keys,
``storage`` and ``types``.  The ``storage`` object is an array where each
element has the following form:

.. code::

    {

>         "astId": 2,

Reviewer: Changed these to be numbers again and added some spaces.

- ``label`` is the name of the state variable
- ``offset`` is the offset in bytes within the storage slot according to the encoding
- ``slot`` is the storage slot where the state variable resides or starts. This
  number may be very large and therefore its JSON value is represented as a
  string.
- ``type`` is an identifier used as key to the variable's type information (described in the following)

The given ``type``, in this case ``t_uint256`` represents an element in
``types``, which has the form:

.. code::

    {
        "encoding": "inplace",
        "label": "uint256",
        "numberOfBytes": "32",
    }

where

- ``encoding`` how the data is encoded in storage, where the possible values are:

  - ``inplace``: data is laid out contiguously in storage (see :ref:`above <storage-inplace-encoding>`).
  - ``mapping``: Keccak-256 hash-based method (see :ref:`above <storage-hashed-encoding>`).
  - ``dynamic_array``: Keccak-256 hash-based method (see :ref:`above <storage-hashed-encoding>`).
  - ``bytes``: single slot or Keccak-256 hash-based depending on the data size (see :ref:`above <bytes-and-string>`).

- ``label`` is the canonical type name.

> - ``numberOfBytes`` is the number of used bytes (as a decimal string). Note that if ``numberOfBytes > 32`` this means that more than one slot is used.

Reviewer: Added string clarification.

        },
        {
          "ast_id": "31",
          "contract": "A",
          "label": "s1",
          "offset": "0",
          "slot": "9",
          "type": "t_string_storage"
        },
        {
          "ast_id": "33",
          "contract": "A",
          "label": "b1",
          "offset": "0",
          "slot": "10",
          "type": "t_bytes_storage"
        }
      ],
      "types": {
        "t_address": {
          "encoding": "inplace",
          "label": "address",
          "numberOfBytes": "20"
        },
        "t_array(t_uint256)2_storage": {
          "base": "t_uint256",
          "encoding": "inplace",
          "label": "uint256[2]",
          "numberOfBytes": "64"
        },
        "t_array(t_uint256)dyn_storage": {
          "base": "t_uint256",
          "encoding": "dynamic_array",
          "label": "uint256[]",
          "numberOfBytes": "32"
        },
        "t_bool": {
          "encoding": "inplace",
          "label": "bool",
          "numberOfBytes": "1"
        },
        "t_bytes1": {
          "encoding": "inplace",
          "label": "bytes1",
          "numberOfBytes": "1"
        },
        "t_bytes_storage": {

>           "base": "t_bytes1",

Reviewer 2: ...

          "encoding": "string",
          "label": "bytes",
          "numberOfBytes": "32"
        },
        "t_mapping(t_address,t_bool)": {
          "encoding": "mapping",
          "key": "t_address",
          "label": "mapping(address => bool)",
          "numberOfBytes": "32",
          "value": "t_bool"
        },
        "t_mapping(t_uint256,t_mapping(t_address,t_bool))": {
          "encoding": "mapping",
          "key": "t_uint256",
          "label": "mapping(uint256 => mapping(address => bool))",
          "numberOfBytes": "32",
          "value": "t_mapping(t_address,t_bool)"
        },
        "t_string_storage": {

>           "base": "t_bytes1",

Reviewer 2: ...
